### PR TITLE
Preserve metric types in top_metrics

### PR DESCRIPTION
**Pull request by nik9000**
This changes the `top_metrics` aggregation to return metrics in their
original type. Since it only supports numerics, that means that dates,
longs, and doubles will come back as stored, with their appropriate
formatter applied.

### CODE REVIEW

File: x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/topmetrics/InternalTopMetrics.java
Comment (nik9000) [inline x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/topmetrics/InternalTopMetrics.java]: Again, we haven't released this so this break is ok.

File: x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/topmetrics/InternalTopMetrics.java
Comment (nik9000) [inline x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/topmetrics/InternalTopMetrics.java]: I'm aware it is odd to have a "SortValue" be part of a MetricValue. It is a very convenient way to send a type-aware thing across the wire though.

Comment (not-napoleon) [reply x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/topmetrics/InternalTopMetrics.java]: I think this would be better as an inline comment than a PR comment.

File: x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/topmetrics/InternalTopMetrics.java
Comment (not-napoleon) [inline x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/topmetrics/InternalTopMetrics.java]:
```suggestion
            return "MetricValue(" + format + "," + value + ")";
```
Pet peeve of mine, unless we have a good reason not to, we should include the class name in the `toString`.  I've gotten burned by this a few times trying to write exception messages.

File: x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/topmetrics/TopMetricsAggregator.java
Comment (not-napoleon) [inline x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/topmetrics/TopMetricsAggregator.java]:
```suggestion
         * that we can steal from the longs to represent missing that
```

**Review by not-napoleon — APPROVED** [review]
LGTM, left a couple of small nits.

File: x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/topmetrics/InternalTopMetrics.java
Comment (nik9000) [reply x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/topmetrics/InternalTopMetrics.java]: :+1:

**Comment by nik9000** [comment]
Thanks @not-napoleon! I think @polyfractal said he was looking at this so I'll probably give him some more time before merging.

File: x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/topmetrics/TopMetricsAggregator.java
Comment (polyfractal) [inline x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/topmetrics/TopMetricsAggregator.java]: Should we put an info blurb somewhere in the docs that multi-valued fields will be averaged together?

**Review by polyfractal — APPROVED** [review]
Had a quick skim, LGTM2.

We might want to add the new "supported type" tests at some point... but since it was designed for single-type aggs it'll probably need some fiddling and I don't want to hold up this PR for that.  :+1:

File: x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/topmetrics/TopMetricsAggregator.java
Comment (nik9000) [reply x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/topmetrics/TopMetricsAggregator.java]: Probably! I'd planned to give the docs another pass in the next PR.

**Pull request merged**
